metals/src/main/scala/scala/meta/internal/bsp/BspConnector.scala

```diff
@@ -0,0 +1,200 @@
+import scala.meta.internal.metals.UserConfiguration
+import scala.meta.io.AbsolutePath
+
+import ch.epfl.scala.bsp4j.BspConnectionDetails
+import com.google.common.collect.ImmutableList
+import org.eclipse.lsp4j.services.LanguageClient
+
+object BspConnector {
+  final val BLOOP_SELECTED = "BLOOP"
+}
+
+class BspConnector(
+    bloopServers: BloopServers,
+    sbtServer: SbtServer,
+    bspServers: BspServers,
+    buildTools: BuildTools,
+    client: LanguageClient,
+    tables: Tables,
+    userConfig: () => UserConfiguration
+)(implicit ec: ExecutionContext) {
+
+  def resolve(): BspResolvedResult = {
+    resolveExplicit().getOrElse {
+      if (buildTools.isBloop) ResolvedBloop
+      else bspServers.resolve()
+    }
+  }
+
+  private def resolveExplicit(): Option[BspResolvedResult] = {
+    tables.buildServers.selectedServer().flatMap { sel =>
+      if (sel == BspConnector.BLOOP_SELECTED) Some(ResolvedBloop)
+      else
+        bspServers
+          .findAvailableServers()
+          .find(_.getName == sel)
+          .map(x => ResolvedBspOne(x))
+    }
+  }
+
+  def connect(
+      workspace: AbsolutePath,
+      userConfiguration: UserConfiguration
+  )(implicit ec: ExecutionContext): Future[Option[BspSession]] = {
+    def connect(
+        workspace: AbsolutePath
+    ): Future[Option[BuildServerConnection]] = {
+      scribe.info("Attempting to connect to the build server...")
+      resolve() match {
+        case ResolvedNone =>
+          scribe.info("No build server found")
+          Future.successful(None)
+        case ResolvedBloop =>
+          scribe.info(
+            "Attempting to start a bloop connection from previous choice"
+          )
+          bloopServers.newServer(workspace, userConfiguration).map(Some(_))
+        case ResolvedBspOne(details) if details.getName == SbtBuildTool.name =>
+          scribe.info(
+            s"Attempting to start a new connection to ${details.getName()} from previous choice..."
+          )
+          // NOTE: we explicity start another sbt process here as simply using newServer
+          // here wil indeed start sbt, but not correctly star the bsp sessions. So before
+          // we start the session we ensure that sbt is running, and then connect.
+          val (_, processHandler) = sbtServer.runSbtShell()
+          processHandler.initialized.future.flatMap { _ =>
+            scribe.info(
+              s"sbt up and running, attempting to start a bsp session..."
+            )
+            bspServers.newServer(workspace, details).map(Some(_))
+          }
+        case ResolvedBspOne(details) =>
+          scribe.info(
+            s"Attempting to start a new connection to ${details.getName()} from previous choice..."
+          )
+          bspServers.newServer(workspace, details).map(Some(_))
+        case ResolvedMultiple(_, _) => Future.successful(None)
+      }
+    }
+
+    // TODO-BSP ensure the timing of this is correct, we don't want the metabuilds yet for sbt BSP
```

Comment:
That's right, we meta builds are currently not supported, so we should most likely skip them. Especially that maybe later on they will be provided in the main connection, which would mean that we don't need to create any additional configuration.

Reply:
Alright, cool. Thanks for confirming 👍